src/libraries/System.DirectoryServices.Protocols/src/System/DirectoryServices/Protocols/common/DirectoryControl.cs

```diff
@@ -699,101 +892,73 @@ public SortKey[] SortKeys
+                // with a maximum length of 23 characters - within the attribute name stack allocation
+                // threshold.
+                Span<byte> scratchSpace = stackalloc byte[AttributeNameStackAllocationThreshold];
 
-            try
-            {
-                void** pMemHandle = (void**)memHandle;
-                IntPtr sortPtr = IntPtr.Zero;
-                int i = 0;
-                for (i = 0; i < keyCount; i++)
+                for (int i = 0; i < _keys.Length; i++)
                 {
-                    sortPtr = Marshal.AllocHGlobal(structSize);
-                    Marshal.StructureToPtr(nativeSortKeys[i], sortPtr, false);
-                    pMemHandle[i] = (void*)sortPtr;
-                }
-                pMemHandle[i] = null;
-
-                bool critical = IsCritical;
-                int error = LdapPal.CreateDirectorySortControl(UtilityHandle.GetHandle(), memHandle, critical ? (byte)1 : (byte)0, ref control);
+                    SortKey key = _keys[i];
 
-                if (error != 0)
-                {
-                    if (LdapErrorMappings.IsLdapError(error))
-                    {
-                        string errorMessage = LdapErrorMappings.MapResultCode(error);
-                        throw new LdapException(error, errorMessage);
-                    }
-                    else
+                    using (writer.PushSequence())
                     {
-                        throw new LdapException(error);
-                    }
-                }
+                        if (!string.IsNullOrEmpty(key.AttributeName))
+                        {
+                            int octetStringLength = s_utf8Encoding.GetByteCount(key.AttributeName);
+                            Span<byte> tmpValue = octetStringLength <= AttributeNameStackAllocationThreshold ? scratchSpace.Slice(0, octetStringLength) : new byte[octetStringLength];
 
-                LdapControl managedControl = new LdapControl();
-                Marshal.PtrToStructure(control, managedControl);
-                BerVal value = managedControl.ldctl_value;
-                // reinitialize the value
-                _directoryControlValue = null;
-                if (value != null)
-                {
-                    _directoryControlValue = new byte[value.bv_len];
-                    Marshal.Copy(value.bv_val, _directoryControlValue, 0, value.bv_len);
-                }
-            }
-            finally
-            {
-                if (control != IntPtr.Zero)
-                {
-                    LdapPal.FreeDirectoryControl(control);
-                }
+                            s_utf8Encoding.GetBytes(key.AttributeName, tmpValue);
+                            writer.WriteOctetString(tmpValue);
+                        }
+                        else
+                        {
+                            writer.WriteOctetString([]);
+                        }
```

Comment:
This if-not-empty { count bytes; make space; UTF-8 encode; write octet string } else { write empty octet string } appears multiple times.  It should be factored out to `internal static void WriteLDAPString(AsnWriter writer, string value)`.  or something like that.